### PR TITLE
Django 1.8

### PR DESCRIPTION
**Pull request by rowanseymour**

### CODE REVIEW

File: pip-freeze.txt [hunk 1]
```diff
@@ -23,7 +23,7 @@ xlutils==1.7.1
 phonenumbers==7.1.0
 django-countries==3.0.1
 djangorestframework==2.4.5
-django-ttag==2.4
+-e git+https://github.com/caktus/django-ttag.git@40c97c01325d436f5d04529526fa7468ff778137#egg=django-ttag-dev
```
Comment (rowanseymour) [inline pip-freeze.txt]: https://github.com/lincolnloop/django-ttag/pull/13

File: temba/flows/models.py [hunk 2]
```diff
@@ -2304,7 +2304,7 @@ def get_step_type(dest, rulesets, actionsets):
             # note that badness happened
             import logging
             logger = logging.getLogger(__name__)
-            logger.exception(unicode(e), exc_info=True)
+            logger.exception(unicode(e))
```
Comment (rowanseymour) [inline temba/flows/models.py]: `exc_info` isn't an allowed kwarg for Python versions < 2.7.7 - it's not needed anyway

File: temba/flows/models.py [hunk 3]
```diff
@@ -2326,7 +2326,7 @@ class RuleSet(models.Model):
     TYPE_CONTACT_FIELD = 'contact_field'
     TYPE_EXPRESSION = 'expression'
 
-    TYPE_WAIT = [TYPE_WAIT_MESSAGE, TYPE_WAIT_RECORDING, TYPE_WAIT_DIGIT, TYPE_WAIT_DIGITS]
+    TYPE_WAIT = (TYPE_WAIT_MESSAGE, TYPE_WAIT_RECORDING, TYPE_WAIT_DIGIT, TYPE_WAIT_DIGITS)
```
Comment (rowanseymour) [inline temba/flows/models.py]: Not related but better for this to be an immutable tuple than a list

File: temba/flows/models.py [hunk 4]
```diff
@@ -3447,13 +3450,9 @@ class FlowStep(models.Model):
     left_on = models.DateTimeField(null=True, db_index=True,
                                    help_text=_("When the user left this step in the flow"))
 
-    messages = models.ManyToManyField(Msg,
-                                      null=True,
```
Comment (rowanseymour) [inline temba/flows/models.py]: `null=True` on a `ManyToManyField` now gives a warning

File: temba/orgs/tests.py [hunk 5]
```diff
@@ -202,7 +202,7 @@ def test_webhook_headers(self):
 
         # check that our webhook settings have changed
         org = Org.objects.get(pk=self.org.pk)
-        self.assertEquals('http://webhooks.uniceflabs.org/', org.get_webhook_url())
+        self.assertEquals('http://webhooks.uniceflabs.org', org.get_webhook_url())
```
Comment (rowanseymour) [inline temba/orgs/tests.py]: URLField no longer adds trailing slash: https://code.djangoproject.com/ticket/22114

File: temba/flows/models.py (hunk 4 above)
Comment (norkans7) [inline temba/flows/models.py]: no migration needed on this?

**Comment by norkans7** [comment]
Looks good

File: temba/flows/models.py (hunk 4 above)
Comment (rowanseymour) [inline temba/flows/models.py]: Oh good call. Figured these params we're being ignored but looks like they still require migration...

File: temba/api/models.py [hunk 6]
```diff
@@ -477,10 +477,14 @@ class APIToken(models.Model):
     Our API token, ties in orgs
     """
     key = models.CharField(max_length=40, primary_key=True)
+
     user = models.ForeignKey(User, related_name='api_tokens')
+
     org = models.ForeignKey(Org, related_name='api_tokens')
+
     created = models.DateTimeField(auto_now_add=True)
-    role = models.ForeignKey(Group, null=True)
```
Comment (rowanseymour) [inline temba/api/models.py]: There already is a migration to remove the `null=True` so assuming this being on the model is an oversight

File: temba/campaigns/tasks.py [hunk 7]
```diff
@@ -28,11 +28,12 @@ def check_campaigns_task(sched_id=None):
     if not r.get(key):
         with r.lock(key, timeout=3600):
             # for each that needs to be fired
-            for fire in EventFire.objects.filter(fired=None, scheduled__lte=timezone.now()).select_related('contact', 'contact.org'):
```
Comment (rowanseymour) [inline temba/campaigns/tasks.py]: Good that Django now errors on invalid field references

File: temba/channels/migrations/0027_auto_20151103_1014.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+    dependencies = [
+        ('channels', '0026_channel_scheme'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='channel',
+            name='alert_email',
+            field=models.EmailField(help_text='We will send email alerts to this address if experiencing issues sending', max_length=254, null=True, verbose_name='Alert Email', blank=True),
```
Comment (rowanseymour) [inline temba/channels/migrations/0027_auto_20151103_1014.py]: Default `max_length` of `EmailField` is now 254

File: temba/contacts/migrations/0027_auto_20151103_1014.py [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+    dependencies = [
+        ('contacts', '0026_add_org_non_active_index'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='contacturn',
+            name='urn',
+            field=models.CharField(help_text='The Universal Resource Name as a string. ex: tel:+250788383383', max_length=255, choices=[('tel', 'Phone number'), ('twitter', 'Twitter handle'), ('ext', 'External identifier')]),
```
Comment (rowanseymour) [inline temba/contacts/migrations/0027_auto_20151103_1014.py]: Missing migration from when we added external scheme

File: temba/orgs/migrations/0011_auto_20151103_1014.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+        ('orgs', '0010_auto_20151002_1417'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='invitation',
+            name='email',
+            field=models.EmailField(help_text='The email to which we send the invitation of the viewer', max_length=254, verbose_name='Email'),
+        ),
+        migrations.AlterField(
```
Comment (rowanseymour) [inline temba/orgs/migrations/0011_auto_20151103_1014.py]: Missing migration from surveyor changes

File: temba/orgs/views.py [hunk 11]
```diff
@@ -451,7 +455,7 @@ def form_valid(self, form):
 
             return super(OrgCRUDL.Import, self).form_valid(form)
 
-    class Export(InferOrgMixin, OrgPermsMixin, SmartFormView):
+    class Export(InferOrgMixin, OrgPermsMixin, SmartTemplateView):
```
Comment (rowanseymour) [inline temba/orgs/views.py]: So that it doesn't try to create a form which then blows up because `fields` isn't being provided

**Comment by rowanseymour** [comment]
(I think) because guardian tables to have been created before we were using migrations, you have to run `./manage.py migrate guardian --fake` before the migrations in this PR

File: temba/campaigns/tasks.py [hunk 12]
```diff
@@ -28,11 +28,12 @@ def check_campaigns_task(sched_id=None):
...
+            for fire in EventFire.objects.filter(fired=None,
+                                                 scheduled__lte=timezone.now()).select_related('contact', 'contact__org'):
                 try:
                     push_task(fire.contact.org, HANDLER_QUEUE, HANDLE_EVENT_TASK, dict(type=FIRE_EVENT, id=fire.id))
 
-                except:  # pragma: no cover
+                except Exception:  # pragma: no cover
                     logger.error("Error running campaign event: %s" % fire.pk, exc_info=True)
```
Comment (norkans7) [inline temba/campaigns/tasks.py]: exc_info here

Comment (rowanseymour) [inline temba/campaigns/tasks.py]: `exc_info` is a valid param for `logger.error`, just not `logger.exception` (in some python versions)

Comment (norkans7) [inline temba/campaigns/tasks.py]: Alright

**Comment by ericnewcomer** [comment]
I branched off this for my current work and have been using it for a couple days. So far so good.

File: temba/contacts/models.py [hunk 13]
```diff
@@ -1539,7 +1539,7 @@ def update_groups_for_contact(cls, contact, field=None):
 
         group_change = False
 
-        for group in ContactGroup.user_groups.filter(**qs_args).exclude(query=None).select_related("contacts"):
+        for group in ContactGroup.user_groups.filter(**qs_args).exclude(query=None).prefetch_related("contacts"):
```
Comment (nicpottier) [inline temba/contacts/models.py]: So assuming this is different now, isn't this going to be insane on the U-Reporters group?

Comment (nicpottier) [inline temba/contacts/models.py]: And why do we want the prefetch?

**Pull request merged**